feat: add permit2 on superchainweth

0xDiscotech opened the pull request:
`just pre-pr` and `just test` were executed successfully.

Possible description
----

**Description**

Added Permit2 functionality to the `SuperchainWETH` contract.

Changes made to `WETH98` to support this:
* Updated the visibility of the `allowance` mapping to `internal` and created a public getter.
* Updated the visibility of the `balanceOf` mapping to `internal` and created a public getter for consistency.

In `SuperchainWETH`:
* Override the `allowance` getter to return the maximum allowance when using Permit2.

**Tests**

Added tests to cover the Permit2 integration with `SuperchainWETH`.


------


Previous Notes: 
* For consistency, on read operations I used the `balanceOf()` and `allowance()` getters, and only used the internal mappings for write operations.
* `balanceOf()` getter is not virtual

CLOSES OPT-278

agusduha committed:
feat: add superchain erc20 bridge (#61)

* feat: add superchain erc20 bridge

* fix: interfaces and versions

0xDiscotech committed:
refactor: optimism superchain erc20 redesign (#62)

* refactor: use oz upgradeable erc20 as dependency

* chore: update interfaces

* fix: tests based on changes

* refactor: remove op as dependency

* feat: add check for supererc20 bridge on modifier

* chore: update tests and interfaces

* chore: update stack vars name on test

* chore: remove empty gitmodules file

* chore: update superchain weth errors

agusduha committed:
test: add superchain erc20 bridge tests (#65)

* test: add superchain erc20 bridge tests

* test: add optimism superchain erc20 beacon tests

* test: remove unnecessary test

* test: tests fixes

* test: tests fixes

0xDiscotech committed:
chore: update missing bridge on natspec (#69)

* chore: update missing bridge on natspec

* fix: natspecs

---------

Co-authored-by: agusduha <agusnduha@gmail.com>

agusduha committed:
fix: remove superchain erc20 base (#70)

0xDiscotech committed:
refactor: update isuperchainweth (#71)


---------

Co-authored-by: agusduha <agusnduha@gmail.com>

agusduha committed:
feat: rename mint/burn and add SuperchainERC20 (#74)

* refactor: rename mint and burn functions on superchain erc20

* chore: rename optimism superchain erc20 to superchain erc20

* feat: create optimism superchain erc20 contract

* chore: update natspec and errors

* fix: superchain erc20 tests

* refactor: make superchain erc20 abstract

* refactor: move storage and erc20 metadata functions to implementation

* chore: update interfaces

* chore: update superchain erc20 events

* fix: tests

* fix: natspecs

* fix: add semmver lock and snapshots

* fix: remove unused imports

* fix: natspecs

---------

Co-authored-by: 0xDiscotech <131301107+0xDiscotech@users.noreply.github.com>

agusduha committed:
fix: refactor zero check (#76)

agusduha committed:
Merge branch 'develop' into fix/redesign-merge-conflicts

agusduha committed:
fix: pre pr

agusduha committed:
Merge pull request #77 from defi-wonderland/fix/redesign-merge-conflicts

fix: update redesign with dev

agusduha committed:
fix: semver natspec check failure (#79)

* fix: semver natspec check failure

* fix: ignore mock contracts in semver natspec script

* fix: error message

agusduha committed:
feat: add crosschain erc20 interface (#80)

* feat: add crosschain erc20 interface

* fix: refactor interfaces

agusduha committed:
fix: superchain bridge natspec (#83)

agusduha committed:
fix: superchain weth natspec (#84)

Co-authored-by: 0xng <ng@defi.sucks>
Co-authored-by: 0xParticle <particle@defi.sucks>
Co-authored-by: gotzenx <78360669+gotzenx@users.noreply.github.com>

agusduha committed:
fix: stop inheriting superchain interfaces (#85)

* fix: stop inheriting superchain interfaces

* fix: move events and erros into the implementation

* fix: make superchainERC20 inherits from crosschainERC20

agusduha committed:
fix: superchain bridge rename (#86)

agusduha committed:
fix: fee vault compiler error (#87)

agusduha committed:
Merge branch 'develop' into fix/crosschain-merge-conflicts

agusduha committed:
fix: remove unused imports

agusduha committed:
Merge pull request #88 from defi-wonderland/fix/crosschain-merge-conflicts

fix: crosschain merge conflicts

agusduha committed:
fix: refactor common errors (#90)

* fix: refactor common errors

* fix: remove unused version

agusduha committed:
fix: reuse unauthorized error (#92)

agusduha committed:
Merge branch 'develop' into fix/superchain-erc20-factory-merge-conflicts

agusduha committed:
fix: superchain erc20 factory conflicts

agusduha committed:
Merge pull request #93 from defi-wonderland/fix/superchain-erc20-factory-merge-conflicts

fix: superchain erc20 factory merge conflicts

agusduha committed:
fix: rename crosschain functions (#94)

0xDiscotech committed:
feat: add permit2 infinite allowance on superchainweth

* refactor: add allowance and balance of getters on weth98

* test: create unit test for the new functionalities

0xDiscotech committed:
chore: run pre-pr

0xDiscotech committed:
chore: remove unnecessary assume

0xng commented on packages/contracts-bedrock/src/L2/WETH.sol:
```diff
@@ -14,8 +14,8 @@ import { IL1Block } from "src/L2/interfaces/IL1Block.sol";
 ///        Allows for nice rendering of token names for chains using custom gas token.
 contract WETH is WETH98, ISemver {
-    /// @custom:semver 1.1.0-beta.2
-    string public constant version = "1.1.0-beta.2";
```
why are we changing the version of this if we are not changing anything?

0xDiscotech replied:
We're modifying the contract it inherits, that produces changes on the `WETH`. AFAIK if I don't update the semver and there are changes on the contract, the `just pre-pr` will fail when it checks that no changes were found on contracts without any version modification

0xng replied:
perfect, missed this inherits from `WETH98` thought it was the other way around

0xng commented on packages/contracts-bedrock/src/universal/WETH98.sol:
```diff
@@ -49,16 +49,26 @@ contract WETH98 is IWETH {
     function withdraw(uint256 wad) public virtual {
-        require(balanceOf[msg.sender] >= wad);
-        balanceOf[msg.sender] -= wad;
+        require(balanceOf(msg.sender) >= wad);
```
i would not make a function call in these cases, i would rather just use the balance to avoid the extra jump. If someone wants to edit the balance of a specific user, that can be accomplished in the constructor of the inheriting contract - the key difference here is that `balanceOf` is the balance of a single user for the entire contract, while `allowance` is the allowance of a user for a specific owner, so to make a global allowance is not as straight forward as updating something in the constructor as it is with `balanceOf`

0xDiscotech replied:
perfect, will use the internal then 👍

0xng commented on packages/contracts-bedrock/src/universal/WETH98.sol:
```diff
@@ -26,8 +26,8 @@ import { IWETH } from "src/universal/interfaces/IWETH.sol";
 
-    mapping(address => uint256) public balanceOf;
-    mapping(address => mapping(address => uint256)) public allowance;
+    mapping(address => uint256) internal _balanceOf;
```
does evertything break if we make both of these `private` over `internal` to follow OZ's style?

0xDiscotech replied:
Yes it fails bc on `SuperchainWETH` we need to access the mappings to perform write operations (on `_mint()` and `_burn()`). With `private` it doesn't find them

agusduha commented on packages/contracts-bedrock/src/universal/WETH98.sol:
```diff
@@ -82,15 +92,16 @@ contract WETH98 is IWETH {
     /// @inheritdoc IWETH
     function transferFrom(address src, address dst, uint256 wad) public returns (bool) {
-        require(balanceOf[src] >= wad);
+        require(balanceOf(src) >= wad);
```
maybe i would use `_balanceOf` here

0xDiscotech committed:
refactor: use internal balance of mapping instead of getter

0xDiscotech committed:
fix: improve flaky test

0xDiscotech committed:
Merge branch 'sc-feat/superweth-permit2' into feat/superweth-permit2

0xDiscotech committed:
chore: run pre-pr

sebastianst committed:
op-node/rollup/derive: Implement Holocene Batch Stage (#12417)

mslipper committed:
op-deployer: Run validations over L2 genesis (#12560)

* op-deployer: Run validations over L2 genesis

Asserts that the L2 genesis is valid using our pre-existing checks prior to running the script.

* add negative test

* fix broken test

* clarify test name

mslipper committed:
op-deployer: Add deploy config inspect command (#12570)

* op-deployer: Add deploy config inspect command

* goimports

* error check

Commit:
e2e test for Holocene's configurable eip-1559 params (#12499)

agusduha commented on packages/contracts-bedrock/test/L2/SuperchainWETH.t.sol:
```diff
@@ -404,7 +404,9 @@ contract SuperchainWETH_Test is CommonTest {
         // Assert
-        assertEq(superchainWeth.balanceOf(_src), 0);
         assertEq(superchainWeth.balanceOf(_dst), _wad);
+        // Handle the case where the source and destination are the same to check the source balance.
+        if (_src != _dst) assertEq(superchainWeth.balanceOf(_src), 0);
+        else assertEq(superchainWeth.balanceOf(_src), _wad);
```
isnt it cleaner to assume `src != dst` ?

I know we are ignoring that explicit case but idk if it matters, no strong opinion

0xDiscotech replied:
To be consistent with how are we doing, would it be better to test that scenario on another test? I think it's valuable to cover that case as well

agusduha replied:
yeah better another test!

agusduha commented on packages/contracts-bedrock/src/L2/interfaces/ISuperchainWETH.sol:
```diff
@@ -11,3 +11,5 @@ interface ISuperchainWETH is IWETH, ICrosschainERC20, ISemver {
     function __constructor__() external;
 }
+
+interface ISuperchainWETHERC20 is IWETH, ISuperchainWETH { }
```
This interface should not exist anymore, it was deleted in other PR

0xDiscotech replied:
perfect

agusduha commented on packages/contracts-bedrock/src/L2/SuperchainWETH.sol:
```diff
@@ -37,19 +38,25 @@ contract SuperchainWETH is WETH98, ICrosschainERC20, ISemver {
     }
 
+    /// @inheritdoc WETH98
+    function allowance(address src, address guy) public view override returns (uint256) {
```
should we replace `src` for `owner` and `guy` for `spender` as in `IWETH`? 

it feels more consistent in this contract

agusduha commented on packages/contracts-bedrock/test/setup/Setup.sol:
```diff
@@ -106,7 +106,7 @@ contract Setup {
     ILegacyMessagePasser legacyMessagePasser = ILegacyMessagePasser(Predeploys.LEGACY_MESSAGE_PASSER);
     IWETH weth = IWETH(payable(Predeploys.WETH));
-    ISuperchainWETH superchainWeth = ISuperchainWETH(payable(Predeploys.SUPERCHAIN_WETH));
+    ISuperchainWETHERC20 superchainWeth = ISuperchainWETHERC20(payable(Predeploys.SUPERCHAIN_WETH));
```
Same here, `ISuperchainWETH` should be the only interface left

agusduha commented on packages/contracts-bedrock/src/universal/WETH98.sol:
```diff
@@ -49,16 +49,26 @@ contract WETH98 is IWETH {
     }
 
+    /// @inheritdoc IWETH
+    function allowance(address src, address guy) public view virtual returns (uint256) {
```
Same here, `owner` and `spender`?

0xDiscotech committed:
fix: address feedback

0xDiscotech committed:
chore: run pre-pr

agusduha reviewed (dismissed):
lgtm

mslipper committed:
op-deployer: Fix standard artifacts urls (#12575)

GrapeBaBa committed:
feat:add cannon mips mul/div tests (#12459)

Signed-off-by: Chen Kai <281165273grape@gmail.com>

ajsutton committed:
rollup-client: Increase call timeouts in CI (#12561)

Refactors the op-service/client package a fair bit to ensure config options are more consistently applied.

0xDiscotech committed:
feat: support permit2 on OptimismSuperchainERC20 + upgrade solady's erc20 implementation (#12476)

* feat: add superchain erc20 bridge (#61)

* feat: add superchain erc20 bridge

* fix: interfaces and versions

* refactor: optimism superchain erc20 redesign (#62)

* refactor: use oz upgradeable erc20 as dependency

* chore: update interfaces

* fix: tests based on changes

* refactor: remove op as dependency

* feat: add check for supererc20 bridge on modifier

* chore: update tests and interfaces

* chore: update stack vars name on test

* chore: remove empty gitmodules file

* chore: update superchain weth errors

* test: add superchain erc20 bridge tests (#65)

* test: add superchain erc20 bridge tests

* test: add optimism superchain erc20 beacon tests

* test: remove unnecessary test

* test: tests fixes

* test: tests fixes

* chore: update missing bridge on natspec (#69)

* chore: update missing bridge on natspec

* fix: natspecs

---------

Co-authored-by: agusduha <agusnduha@gmail.com>

* fix: remove superchain erc20 base (#70)

* refactor: update isuperchainweth (#71)


---------

Co-authored-by: agusduha <agusnduha@gmail.com>

* feat: rename mint/burn and add SuperchainERC20 (#74)

* refactor: rename mint and burn functions on superchain erc20

* chore: rename optimism superchain erc20 to superchain erc20

* feat: create optimism superchain erc20 contract

* chore: update natspec and errors

* fix: superchain erc20 tests

* refactor: make superchain erc20 abstract

* refactor: move storage and erc20 metadata functions to implementation

* chore: update interfaces

* chore: update superchain erc20 events

* fix: tests

* fix: natspecs

* fix: add semmver lock and snapshots

* fix: remove unused imports

* fix: natspecs

---------

Co-authored-by: 0xDiscotech <131301107+0xDiscotech@users.noreply.github.com>

* fix: refactor zero check (#76)

* fix: pre pr

* chore: add new solady version and import it for erc20

* fix: undo forge std changes

* chore: re run pre pr script

* fix: semver natspec check failure (#79)

* fix: semver natspec check failure

* fix: ignore mock contracts in semver natspec script

* fix: error message

* feat: add crosschain erc20 interface (#80)

* feat: add crosschain erc20 interface

* fix: refactor interfaces

* fix: superchain bridge natspec (#83)

* fix: superchain weth natspec (#84)

Co-authored-by: 0xng <ng@defi.sucks>
Co-authored-by: 0xParticle <particle@defi.sucks>
Co-authored-by: gotzenx <78360669+gotzenx@users.noreply.github.com>

* fix: stop inheriting superchain interfaces (#85)

* fix: stop inheriting superchain interfaces

* fix: move events and erros into the implementation

* fix: make superchainERC20 inherits from crosschainERC20

* fix: superchain bridge rename (#86)

* fix: fee vault compiler error (#87)

* fix: remove unused imports

* chore: run pre-pr and update vendor interface

* fix: refactor common errors (#90)

* fix: refactor common errors

* fix: remove unused version

* feat: add permit2 on optimism superchain erc20

* chore: run pre-pr script

* fix: reuse unauthorized error (#92)

* fix: superchain erc20 factory conflicts

* fix: rename crosschain functions (#94)

* chore: run pre-pr

* chore: run pre-pr

* chore: run pre-pr

* feat: add new tests on optimism superchain erc20

* fix: vars and params naming on newly added tests

* fix: var name

* feat: support permit2 on optimism superchain erc20 and upgrade solady's erc20 implementation (#97)


---
Co-Authored-by: AgusDuha <81362284+agusduha@users.noreply.github.com>

* chore: use ierc20 alias for ierc20 solady interface (#108)

---------

Co-authored-by: AgusDuha <81362284+agusduha@users.noreply.github.com>
Co-authored-by: agusduha <agusnduha@gmail.com>
Co-authored-by: 0xng <ng@defi.sucks>
Co-authored-by: 0xParticle <particle@defi.sucks>
Co-authored-by: gotzenx <78360669+gotzenx@users.noreply.github.com>

ajsutton committed:
proofs-tools: Update kona (#12581)

mslipper committed:
op-deployer: Add support for inspecting l2 semvers (#12577)

* op-deployer: Add support for inspecting l2 semvers

* goimports

* add factory targets

ajsutton committed:
proofs-tools: Update kona to alpha 5 (#12583)

mslipper committed:
op-service: Add omitempty to EIP1559Params (#12582)

* op-deployer: Add support for inspecting l2 semvers

* goimports

* add factory targets

* op-service: Add omitempty to EIP1559Params

These parameters are only valid for Holocene, and cause errors during unmarshaling on older op-node versions.

ajsutton committed:
op-program: Automatically verify the prestate hash for all op-program tags (#12538)

Generates a versions.json to record which prestate came from each tag and compares it to the releases.json.
Adds a couple of releases missed from releases.json and removes governanceApproved from some releases which were not deployed to op-mainnet or submitted to governance.

ajsutton committed:
op-challenger: Use info log level for kona not default (#12585)

Kona changed the default log level so the -v is undesirable now.

ajsutton committed:
op-challenger: Use binary snapshots for asterisc (#12586)

ajsutton committed:
op-challenger: Log warning and continue if game impl not available (#12539)

0xDiscotech committed:
Merge branch 'develop' into feat/superweth-permit2

0xDiscotech committed:
chore: run pre-pr

0xDiscotech commented:
This PR will be moved to https://github.com/defi-wonderland/optimism/pull/115 since the `sc-feat` base branch was updated with the last changes, so we have a smaller diff.
